### ISSUE TITLE
JSON logging isn't truly JSON logging

### ISSUE THREAD

**Issue by kemra102** [aws/aws-node-termination-handler#249]
We have JSON logging enabled. But the messages are basically plain text (not quite truly Human readable). For example here is the log for a spot termination:

```
Got interruption event from channel {InstanceID:i-0cb9fad8f8e327c4c InstanceType:m5d.xlarge PublicHostname:ec2-54-76-23-97.eu-west-1.compute.amazonaws.com PublicIP:54.76.23.97 LocalHostname:ip-10-200-205-41.aeuw1.env.product.cloud.zone LocalIP:10.200.205.41} {EventID:spot-itn-5ef7cf326c13be727fb52feb7ff7ee5658b36478d4697ab81603a2e727a7bb99 Kind:SPOT_ITN Description:Spot ITN received. Instance will be interrupted at 2020-09-01T09:44:02Z 
 State: StartTime:2020-09-01 09:44:02 +0000 UTC EndTime:0001-01-01 00:00:00 +0000 UTC Drained:false PreDrainTask:0x10cc2f0}
```

This is a string, with a regular string upfront and 2 embedded JSON data structures on the end. As such, automated JSON log parsing in products such as Logstash/ElasticSearch don't work. So even after turning on JSON logging if we wanted to properly parse these log entries we'd still need a custom transformation rule in Logstash/ES to do so. Which kind of defeats the point in my opinion.

This is just one example, other parts of the application don't appear to even try to do JSON logging, such as startup config printing for example.

I think making the entire app "truly" support JSON logging is a huge effort and I'm not sure how to go about it or even if this is desirable to the project maintainers.

**Comment by haugenj** [comment]
Hey @kemra102 !

I agree the JSON logging could be better. When we added JSON logging with Zerolog we wanted to make sure the default human-readable logs displayed in exactly the same format as they were before, in case anyone was scraping the logs and parsing them already. However, like you brought up this limits the JSON log format and keeps it from really being JSON logging. Since that time we've added Prometheus metrics to NTH so changing the format might not be as much of an issue anymore. 

I haven't worked with Logstash or ElasticSearch, so if we did change the format here what would be the ideal output? Anything different than the following?
```
{
"message" : "Got interruption event from channel",
"time" : "2020-09-03T00:00:00Z",
"level" : "info",
"metadata" :
  {
  "InstanceID" : "i-12345678",
  "InstanceType" : "t3.large",
  ...
  },
"event" :
  {
  "EventID" : "spot-itn-asdfghjkl",
  "Kind":"SPOT_ITN",
  ...
  }
}
``` 

The places we're missing JSON logging is a mistake, I could've missed a few spots when I implemented JSON logging and we might not be as vigilant on new PRs as we should be.

**Comment by kemra102** [comment]
Hey @haugenj - thanks for responding so quickly!

At my current workplace we sadly do not have Prometheus, that would actually be my preference. We may have it in the future, but for the time being...

For the purposes of Logstash/Elasticsearch any "pure" JSON would be fine. Logstash would parse it for example in your sample snippet, you would have a field that has a key of `message` that has a value of `Got interruption event from channel` and so on. It basically converts the JSON into an ES document with much the same format as the JSON. This means we could search on things like `time` and even filter right down to something like `event.Kind` = `SPOT_ITN` for example.

**Comment by bwagner5** [comment]
JSON logging enhancements were released in v1.8.0 https://github.com/aws/aws-node-termination-handler/releases/tag/v1.8.0
